src/debug/javaDockerResolver.ts

```diff
@@ -0,0 +1,71 @@
+import * as vscode from 'vscode';
+
+import { IDockerParser, IDockerResolver, PortInfo } from './debugInterfaces';
+import { Kubectl } from '../kubectl';
+import { shell } from '../shell';
+
+const defaultJavaDebugPort = "5005";
+const defaultJavaAppPort = "9000";
+const defaultJavaDebugOpts = `-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=${defaultJavaDebugPort},quiet=y`;
+const javaDebugOptsRegExp = /(-agentlib|-Xrunjdwp):\S*(address=[^\s,]+)/i;
+const fullJavaDebugOptsRegExp = /^java\s+.*(-agentlib|-Xrunjdwp):\S*(address=[^\s,]+)\S*/i;
+
+export class JavaDockerResolver implements IDockerResolver {
+    public isSupportedImage(baseImage: string): boolean {
+        return baseImage.indexOf("java") >= 0
+            || baseImage.indexOf("openjdk") >= 0
+            || baseImage.indexOf("oracle") >= 0;
+    }
+
+    public async resolvePortsFromFile(dockerParser: IDockerParser, env: {}): Promise<PortInfo> {
+        const portInfo: PortInfo = {
+            debug: null,
+            app: null
+        };
+
+        // Resolve the debug port.
+        const matches = dockerParser.searchLaunchArgs(javaDebugOptsRegExp);
+        if (matches) {
+            const addresses = matches[2].split("=")[1].split(":");
+            portInfo.debug = addresses[addresses.length - 1];
+        } else if (dockerParser.searchLaunchArgs(/\$\{?JAVA_OPTS\}?/)) {
+            env["JAVA_OPTS"] = defaultJavaDebugOpts;
+            portInfo.debug = defaultJavaDebugPort;
+        }
+        // Cannot resolve the debug port from Dockerfile, then ask user to specify it.
+        if (!portInfo.debug) {
+            const input = await vscode.window.showInputBox({
+                prompt: `Please specify debug port exposed by the Dockerfile (e.g. 5005)`,
+                placeHolder: "5005"
+            });
+            portInfo.debug = (input ? input.trim() : null);
+        }
+        if (!portInfo.debug) {
+            return portInfo;
+        }
+        
+        // Resolve the app port.
+        const dockerExpose = dockerParser.getExposedPorts();
+        if (portInfo.debug && dockerExpose.length) {
+            const possiblePorts = dockerExpose.filter((port) => port !== portInfo.debug);
+            if (possiblePorts.length === 1) {
+                portInfo.app = possiblePorts[0];
+            } else if (possiblePorts.length > 1) {
+                portInfo.app = await vscode.window.showQuickPick(possiblePorts, { placeHolder: "Please select the app port exposed at Dockerfile" });
+            }
+            // If the exposed port is a variable, then need set it in environment variables.
+            if (/\$\{?(\w+)\}?/.test(portInfo.app)) {
+                const varName = portInfo.app.match(/\$\{?(\w+)\}?/)[1];
+                env[varName] = defaultJavaAppPort;
+                portInfo.app = defaultJavaAppPort;

```

Comment:
Oh so this is why the app port isn't typed as number?  Can we at least type it as `number | string` and document that it can contain an environment variable?  Or better create a `PortNumber` type which can contain a number or an EV name and has well named members and suitable methods for telling which it is?

Reply:
Oh no wait, the portInfo.app is still a number, you just also add an EV that tells the debuggee to use that port number.

Reply:
The original port parsed from dockerfile or regular expression is string type, so i'd like to keep it as string.

Reply:
That doesn't seem like a good reason.  If we followed that logic then everything read from a file or console would be string typed.  Typing it as number communicates what it is to readers of the code.  Even typing it as `number | string` would at least give readers some boundaries on what it might be.  Can you explain the benefit of typing it as `any`?

Reply:
fixed.